pkg/flow/internal/controller/component_node_manager.go

```diff
@@ -0,0 +1,258 @@
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	importNode, exist := m.importNodes[cc.importLabel]
+	return importNode, exist
+}
+
+// CustomComponentConfig represents the config needed by a custom component to load.
+type CustomComponentConfig struct {
+	declareContent            string            // represents the corresponding declare as plain string
+	additionalDeclareContents map[string]string // represents the additional declares that might be needed by the component to build custom components
+}
+
+// NewCustomComponentConfig creates a new CustomComponentConfig.
+func NewCustomComponentConfig(declareContent string, additionalDeclareContents map[string]string) *CustomComponentConfig {
+	return &CustomComponentConfig{
+		declareContent:            declareContent,
+		additionalDeclareContents: additionalDeclareContents,
+	}
+}
+
+// GetCustomComponentConfig returns the custom component config for a given custom component or an error if not found.
+func (m *ComponentNodeManager) GetCustomComponentConfig(cc *CustomComponentNode) (*CustomComponentConfig, error) {
+	m.mut.Lock()
+	defer m.mut.Unlock()
+	var customComponentConfig *CustomComponentConfig
+	var err error
+	if cc.importLabel == "" {
+		customComponentConfig = m.getCustomComponentConfigFromLocalDeclares(cc)
+		if customComponentConfig == nil {
+			customComponentConfig = m.getCustomComponentConfigFromParent(cc)
+		}
+	} else {
+		customComponentConfig, err = m.getCustomComponentConfigFromImportedDeclares(cc)
+		if err != nil {
+			return customComponentConfig, err
+		}
+		if customComponentConfig == nil {
+			customComponentConfig = m.getCustomComponentConfigFromParent(cc)
+			// Custom components that receive their config from imported declares in a parent controller can only access the imported declares coming from the same import.
+			customComponentConfig.additionalDeclareContents = filterAdditionalDeclareContents(cc.importLabel, customComponentConfig.additionalDeclareContents)
+		}
+	}
+	if customComponentConfig == nil {
+		return nil, fmt.Errorf("custom component config not found for component %s", cc.componentName)
+	}
+	return customComponentConfig, nil
+}
+
+// getCustomComponentConfigFromLocalDeclares retrieves the config of a custom component from the local declares.
+func (m *ComponentNodeManager) getCustomComponentConfigFromLocalDeclares(cc *CustomComponentNode) *CustomComponentConfig {
+	node, exists := m.declareNodes[cc.declareLabel]
+	if !exists {
+		return nil
+	}
+	return NewCustomComponentConfig(node.Declare().content, m.getLocalAdditionalDeclareContents(cc.componentName))
+}
+
+// getCustomComponentConfigFromParent retrieves the config of a custom component from the parent controller.
+func (m *ComponentNodeManager) getCustomComponentConfigFromParent(cc *CustomComponentNode) *CustomComponentConfig {
+	declareContent, exists := m.additionalDeclareContents[cc.componentName]
```

Comment:
Can we elaborate more on this? I am not sure if I follow the use case here. Do children inherit the imports of their parents?

Reply:
That's correct. For example:

```
declare "first" {}
declare "second" {
  first "nested" {} 
}
second "root" {}
```

In this code, the declare "first" is declared at the root but used inside of the declare "second".

When you evaluate "root", you will create a new custom component and load a new config with a new controller. But this controller only has whatever is in "declare "second"" as context. It cannot see what's declared in the parent.

That's why the parent needs to pass the definition of "first" to its child. That would be the same with import.

I will consider adding more comments to clarify these points but first I will pair up with Robert for a review

Reply:
OK, that makes sense, for some reason I was thinking child and parent module relationships and not child and parent node relationships.